pkg/certificate/chain/chain.go

```diff
@@ -0,0 +1,288 @@
+func (r *certificateChain) setCaKeyPair(keyPair *triple.KeyPair) error {
+	r.ca.keyPair = keyPair
+	r.ca.KeyPEM, r.ca.CertPEM = keyPairToKeyPairPem(keyPair)
+	for _, c := range r.certificatesIssued {
+		for k, certs := range c.caCerts {
+			certs = append(certs, keyPair.Cert)
+			c.caCerts[k] = certs
+			c.CACertPEM[k] = triple.EncodeCertsPEM(certs)
+		}
+	}
+	return nil
+}
+
+// setCaCerts sets a CA certificate for a certificate issue in all formats
+func (r *certificateChain) setCaCerts(s *certificateIssue, name string, certs []*x509.Certificate) {
+	s.caCerts[name] = certs
+	s.CACertPEM[name] = triple.EncodeCertsPEM(certs)
+}
+
+// setKeyResetCert sets a key pair for a certificate issue in all formats, existing certificates are removed
+func (r *certificateChain) setKeyResetCert(s *certificateIssue, keyPair *triple.KeyPair) {
+	s.key = keyPair.Key
+	s.certs = []*x509.Certificate{keyPair.Cert}
+	s.KeyPEM, s.CertPEM = keyPairToKeyPairPem(keyPair)
+}
+
+// setKeyAppendCert sets a key pair for a certificate issue in all formats, appended to previous certificates
+func (r *certificateChain) setKeyAppendCert(s *certificateIssue, keyPair *triple.KeyPair) {
+	s.key = keyPair.Key
+	s.certs = append(s.certs, keyPair.Cert)
+	s.KeyPEM = triple.EncodePrivateKeyPEM(keyPair.Key)
+	s.CertPEM = triple.EncodeCertsPEM(s.certs)
+}
+
+// setCerts sets certificates for a certificate issue in all formats, preserving the previous key
+func (r *certificateChain) setCerts(s *certificateIssue, certs []*x509.Certificate) {
+	s.certs = certs
+	s.CertPEM = triple.EncodeCertsPEM(certs)
+}
+
+// keyPairPemToKeypair converts KeyPair from PEM format
+func keyPairPemToKeypair(keypem []byte, certpem []byte) (*rsa.PrivateKey, []*x509.Certificate, error) {
+	key, err := triple.ParsePrivateKeyPEM(keypem)
+	if err != nil {
+		return nil, nil, err
+	}
+	rsakey, ok := key.(*rsa.PrivateKey)
+	if !ok {
+		return nil, nil, errors.New("Expected RSA key but found different type")
+	}
+
+	certs, err := triple.ParseCertsPEM(certpem)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return rsakey, certs, nil
+}
+
+// keyPairToKeyPairPem converts KeyPair to PEM format
+func keyPairToKeyPairPem(keyPair *triple.KeyPair) (key []byte, cert []byte) {
+	key = triple.EncodePrivateKeyPEM(keyPair.Key)
+	cert = triple.EncodeCertPEM(keyPair.Cert)
+	return
+}
+
+func (r *certificateChain) update() (time.Time, error) {
+	logger := r.log.WithName("update")
+	logger.Info("Checking certificate chain for rotation or cleanup")
+
+	deadlineToRotateCA := r.findRotationDeadlineForCA()
+	deadlineToRotateCerts := r.findRotationDeadlineForCerts()
+	rotateCA := !r.now().Before(deadlineToRotateCA)
+	rotateCerts := !r.now().Before(deadlineToRotateCerts)
+
+	// Ensure certificate chain
+	if !rotateCA {
+		err := r.verifyTLS()
+		if err != nil {
+			logger.Info("Certificate chain failed verification, will force full chain rotation", "err", err)
+			// Force rotation
+			rotateCA = true
+		}
+	}
```

Comment:
If a service certificate expire it will enter here and all the certificates will be rotated, we don't need this block.

Reply:
Just as it was being done before, I am not changing functionality here.

Reply:
Right... and now I see it was wrong, let's fix it at follow up PR.